lib/block-supports/index.php

```diff
@@ -65,40 +65,36 @@ function gutenberg_apply_block_supports( $block_content, $block ) {
 		return $block_content;
 	}
 
-	// Some inline styles may be added without ending ';', add this if necessary.
-	$current_styles = trim( $block_root->getAttribute( 'style' ), ' ' );
-	if ( strlen( $current_styles ) > 0 && substr( $current_styles, -1 ) !== ';' ) {
-		$current_styles = $current_styles . ';';
-	};
-
-	// Merge and dedupe new and existing classes and styles.
-	$classes_to_add = esc_attr( implode( ' ', array_key_exists( 'css_classes', $attributes ) ? $attributes['css_classes'] : array() ) );
-	$styles_to_add  = esc_attr( implode( ' ', array_key_exists( 'inline_styles', $attributes ) ? $attributes['inline_styles'] : array() ) );
-	$new_classes    = implode( ' ', array_unique( explode( ' ', ltrim( $block_root->getAttribute( 'class' ) . ' ' ) . $classes_to_add ) ) );
-	// Normalize styles for dedupe.
-	$all_styles        = explode( ';', $current_styles . $styles_to_add );
-	$normalized_styles = array();
-	foreach ( $all_styles as $style ) {
-		if ( '' === $style ) {
-			continue;
-		}
-		// Split key and value of style.
-		$split_style = explode( ':', $style );
-		// Trim whitespace and recombine the style in a consistent form.
-		$style               = trim( $split_style[0] ) . ': ' . trim( $split_style[1] ) . ';';
-		$normalized_styles[] = $style;
-	}
-	$new_styles = implode( ' ', array_unique( $normalized_styles ) );
+	$current_classes = explode( ' ', trim( $block_root->getAttribute( 'class' ) ) );
+	$classes_to_add  = array_key_exists( 'css_classes', $attributes ) ? $attributes['css_classes'] : array();
+	$new_classes     = array_unique( array_filter( array_merge( $current_classes, $classes_to_add ) ) );
+
+	$current_styles = preg_split( '/\s*;\s*/', trim( $block_root->getAttribute( 'style' ) ) );
+	$styles_to_add  = array_key_exists( 'inline_styles', $attributes ) ? $attributes['inline_styles'] : array();
+	$new_styles     = array_unique( array_map( 'gutenberg_normalize_css_rule', array_filter( array_merge( $current_styles, $styles_to_add ) ) ) );
 
 	// Apply new styles and classes.
 	if ( ! empty( $new_classes ) ) {
-		$block_root->setAttribute( 'class', $new_classes );
+		$block_root->setAttribute( 'class', esc_attr( implode( ' ', $new_classes ) ) );
```

Comment:
Im wondering if we should be using `esc_attr` on the new classes and styles before we call `array_unique`.  It is assumed at this point that the incoming classes/styles have been escaped, so we probably want our new ones escaped as well before we dedupe?

Reply:
Hm, I was thinking of `esc_attr` merely from an escaping / safety point of view. If I understood you well, you're suggesting using it also as part of the normalisation? That is, to better catch duplicates between current and new classes?

If so, I'm wondering what that looks like in practice. How likely will there be _legal_ inputs for className such that `esc_attr( s ) !== s`?

Reply:
> I was thinking of esc_attr merely from an escaping / safety point of view.

Yeah, I think thats 99.99% of it.

> If I understood you well, you're suggesting using it also as part of the normalisation? That is, to better catch duplicates between current and new classes?

Essentially.  When I first saw the `esc_attr` this early in the process I assumed that was the reason why.

> How likely will there be legal inputs for className such that esc_attr( s ) !== s?

Probably not very likely.  I can't think of a situation but I lack imagination here. 🤷‍♀️   If you don't think it makes sense either then we can probably leave it until the last step as done here.  If that odd situation does come up, we just end up with a duplicate style rule in the string which isn't the end of the world?

Reply:
>  If that odd situation does come up, we just end up with a duplicate style rule in the string which isn't the end of the world?

Exactly my thinking. :)

Reply:
> Hm, I was thinking of `esc_attr` merely from an escaping / safety point of view. If I understood you well, you're suggesting using it also as part of the normalisation? That is, to better catch duplicates between current and new classes?
> 
> If so, I'm wondering what that looks like in practice. How likely will there be _legal_ inputs for className such that `esc_attr( s ) !== s`?

I feel stupid for asking this question about classes but not to styles. 🤦 (Looking at the fix in #25079.)